Teach red-knot that `type(x)` is the same as `x.__class__`

AlexWaygood opened the pull request:
## Summary

I keep trying to write mdtests that use `type(x)` to get the type of `x`, and expecting red-knot to understand that `type(x)` should be inferred as the meta-type of `x`, the same as `x.__class__`. This PR adds the necessary special-casing so that red-knot will now understand this.

## Test Plan

I added a couple of new assertions, integrated into the existing mdtest for `__class__` inference (since it's two ways of accessing the same type). We're also able to remove an existing TODO in an mdtest for `issubclass()` type narrowing as a result of this change.

AlexWaygood commented:
We already have test coverage for three-argument calls to `type` that would fail if we stopped inferring `builtins.type` for those calls, FWIW:

https://github.com/astral-sh/ruff/blob/b3c5932fda77f3df33ad7a8a90133ca1081b7acc/crates/red_knot_python_semantic/resources/mdtest/narrow/isinstance.md?plain=1#L97-L110

carljm commented on crates/red_knot_python_semantic/resources/mdtest/attributes.md:
```diff
@@ -884,13 +884,18 @@ def _(flag: bool):
 
 ## Objects of all types have a `__class__` method
 
+The type of `x.__class__` is the same as `x`'s meta-type. `x.__class__` is always the same value as
```
"always" is a dangerous word 😆 pretty sure it's possible to break this invariant at runtime with a C extension type. But I also think it's reasonable for us to consider it an invariant.

AlexWaygood replied:
Ugh, really?? I thought that was one of the Big Things that was fixed as part of the 2-to-3 transition 😆 but you probably know better than I!

carljm replied:
It's possible I'm wrong and this is no longer possible for C extension types either! Not going to go down the rabbit hole of experimenting with it to find out!!

carljm commented on crates/red_knot_python_semantic/resources/mdtest/narrow/issubclass.md:
```diff
@@ -112,8 +112,7 @@ def _(flag: bool):
     if issubclass(t, type(None)):
-        # TODO: this should be just `Literal[NoneType]`
-        reveal_type(t)  # revealed: Literal[int, NoneType]
+        reveal_type(t)  # revealed: Literal[NoneType]
```
I love these unexpected TODO fixes, where multiple features compose together and it just... does the right thing

carljm reviewed (approved):
Nice!

AlexWaygood commented:
> We already have test coverage for three-argument calls to `type` that would fail if we stopped inferring `builtins.type` for those calls, FWIW:

I changed my mind about this and added a dedicated test. The pre-existing test only tested the three-argument form of `type()` because of the specific setup the test was using (which actually predated our support for parameter types, anyway). I simplified the existing test that used three-argument `type()` at the same time.
